aten/src/ATen/core/LazyTensor.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <ATen/core/TensorBody.h>
+#include <ATen/core/ivalue.h>
+#include <ATen/core/function_schema.h>
+#include <c10/core/TensorImpl.h>
+
+namespace at {
+
+struct CAFFE2_API LazyTensorImpl : public TensorImpl {
+  explicit LazyTensorImpl(TensorTypeSet ts)
+      : TensorImpl(
+            ts.add(TensorTypeId::LazyTensorId),
+            caffe2::TypeMeta::Make<float>(), // Hack to support gradients
+            c10::Device(c10::DeviceType::CPU)) {}
+
+  explicit LazyTensorImpl(at::Tensor const& inp)
+      : LazyTensorImpl(inp.type_set()) {
+    inps_.emplace_back(inp);
+  }
+
+  explicit LazyTensorImpl(
+      c10::optional<c10::FunctionSchema> schema,
+      std::vector<c10::IValue> const& inps,
+      size_t index = 0)
+      : LazyTensorImpl([&inps]() {
+          TensorTypeSet ts(TensorTypeId::LazyTensorId);
+          for (const auto& inp : inps) {
+            if (inp.isTensor()) {
+              ts = ts | inp.toTensor().type_set();
+            }
+          }
+          return ts;
+        }()) {
+    schema_.swap(schema);
+    for (const auto& inp : inps) {
+      inps_.emplace_back(inp);
+    }
+    index_ = index;
+  }
+
+  c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(
```

Comment:
This method... is weird. CC'ing @yf225. This should only get called if I call `y = x.detach()` on a tensor. I guess because lazy tensor currently lives after autograd, we are indeed obligated to make a shallow copy in this case (so yes, we need to implement it). But the way you are making the shallow copy here seems to imply that you will *redo* the computation in the detached node. That seems wrong: we should reuse the computation result, right?

Reply:
For more context on the semantics: I believe `shallow_copy_and_detach` is only used when we want the new tensor to share the same tensor metadata (e.g. sizes / strides / storage pointer / storage_offset) with the old tensor, but with a new autograd history.

Reply:
honestly I'm not really sure the purpose of this method and cargo-culted it over.

Reply:
the computation will be reused, I believe, by the cache lookup

Reply:
In that case, I think my comments in https://github.com/pytorch/pytorch/pull/30674/files#r356435098 apply